### ISSUE TITLE
Mistral 7B | `RuntimeError: query and key must have the same dtype`

### ISSUE THREAD

**Issue by flexchar** [axolotl-ai-cloud/axolotl#695]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Not fail.

### Current behaviour

It fails after running for around 5.5 minutes. I tried changing settings but no luck. I also tried updating peft library as per this post: https://github.com/OpenAccess-AI-Collective/axolotl/pull/132#issuecomment-1574882573

Terminal log (file names are a little bit different than the rest of the issue but have identical contents).
https://app.warp.dev/block/kqypTiu3ext2XlJPXHIGjY

### Steps to reproduce

Based off the official `config.yaml` for Mistral I built my own and attempt to train using
`accelerate launch -m axolotl.cli.train config.yml`

Running on RunPod A100 using runpod/pytorch:2.0.1-py3.10-cuda11.8.0-devel (the Axolotl docker image always gives issues with pytorch so I don't use it). 

### Config yaml

```yaml
base_model: mistralai/Mistral-7B-v0.1
base_model_config: mistralai/Mistral-7B-v0.1
model_type: MistralForCausalLM
tokenizer_type: LlamaTokenizer
is_mistral_derived_model: true

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: merged.jsonl
    ds_type: json # see other options below
    type: completion
dataset_prepared_path:
val_set_size: 0.01
output_dir: ./lora-out

sequence_len: 8192
sample_packing: true
pad_to_sequence_len: true

wandb_project: abc-project
wandb_entity:
wandb_watch:
wandb_run_id: abc-1-mistral-merged
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 2
num_epochs: 10
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.000005

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
eval_steps: 20
eval_table_size: 5
eval_table_max_new_tokens: 128
save_steps:
debug:
deepspeed:
weight_decay: 0.0
fsdp:
fsdp_config:
special_tokens:
  bos_token: '<s>'
  eos_token: '</s>'
  unk_token: '<unk>'
```

### Possible solution

It worked well with Llama 2 but it's not a solution...

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10.6

### axolotl branch-commit

main/29b8f46aeddf3919034584a0fa7214394510e7c9

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
I assume this is a pretty small dataset. Could you try increasing the val_set_size from 0.01 to 0.05, and also try it with `eval_sample_packing: false`

**Comment by flexchar** [comment]
I can try that. 

Side note: I end up moving to lora with `load_in_8bit: true` since I realised that this is old-fashion/full weights training. That worked. I am curious if `load_in_8bit: false` and no lora means it'd be better/faster to train. The dataset has around 10k json lines/examples. Some of them are a few tokens and others can be all the way up to 3k. 

By the way, can you explain how does the progress bar come up with the total number of iterations? Is it number of epochs (steps) * examples in dataset * something else? I cannot seem to figure out and estimate it.

**Comment by NanoCode012** [comment]
> the Axolotl docker image always gives issues with pytorch so I don't use it

Hey, I noticed this one thing. Could you also please clarify what were the issues? Which image did you use that had the issue?

**Comment by flexchar** [comment]
Sorry Nano, I will try again one day. I think the issue was that I used H100 instead of A100 and it was maybe a driver issue. I then went back to PyTorch and A100. So it's not entirely fair.

---

Thanks @winglian. Perhaps it was it that helped. I retried using this example https://huggingface.co/ehartford/dolphin-2.1-mistral-7b/blob/main/configs/dolphin-mistral-7b.yml and it has your advice implemented for `val_set_size`. I'm gonna close this for now. Pleasure working with this tool!
